subprojects/core/src/main/java/org/gradle/caching/configuration/internal/BuildCacheServiceRegistration.java

```diff
@@ -14,26 +14,12 @@
  * limitations under the License.
  */
 
-package org.gradle.caching;
+package org.gradle.caching.configuration.internal;
 
-import org.gradle.api.Incubating;
+import org.gradle.caching.BuildCacheService;
+import org.gradle.caching.configuration.BuildCache;
 
-/**
- * Factory interface to be provided by build cache service implementations.
- *
- * @param <T> the type of build cache configuration this factory can handle.
- *
- * @since 3.5
- */
-@Incubating
-public interface BuildCacheServiceFactory<T extends org.gradle.caching.configuration.BuildCache> {
-    /**
-     * Returns the type of the supported build cache configuration.
-     */
+public interface BuildCacheServiceRegistration<T extends BuildCache, S extends BuildCacheService> {
     Class<T> getConfigurationType();
```

Comment:
This is only for our internal built-in build cache types, so we can automatically register them as available types.  

Everyone else goes through the `registerBuildCacheService`.  We could make that take 3 types.  An interface that extends `BuildCache`, a class that implements that interface and a class that implements `BuildCacheService`.  I'm sort of OK with a simple POJO configuration class for now and if we decide to do something fancier (like generate an implementation from a configuration interface), we can introduce another method.